### DISCUSSION TITLE
How to toggle a module / set status modules session-specificly?

### DISCUSSION THREAD

**Discussion by twio142** [catppuccin/tmux#257] [Help]
I made a custom module that displays a cloud icon in ssh sessions. But I can't manage to completely hide it when it's not in a ssh session. My `show_ssh` function returns an empty string in that case, but an extra set of status separators still remain in the status bar.

Is there a way to toggle a module in a session? Or can I set the `@catppuccin_status_modules_right` option session-specificly?

**Comment by vdbe** [comment]
If it is a custom module you can do something like 
```bash
show_test() {
  if [ -n "$SSH_TTY" ]; then
    # ...
    module=$( build_status_module "$index" "$icon" "$color" "$text" )

    echo "$module"
  fi
}
```
and not print/echo anything if not in ssh

**Reply by vdbe** [reply]
~~ow forgot about the status separator (can't read apparently) don't think you can get rid of those.~~ No the status separators are added by `build_status_module` which is inside the if.

**Reply by twio142** [reply]
I've tested with the environment variable `$SSH_CONNECTION`, but somehow it's either not accessible in the script (while it's indeed available in the tmux session), or it's been applied globally, affecting all active sessions no matter ssh or not.

My workaround was:

In `~/.tmux.conf`:

```
set -ga update-environment[0] "SSH_CONNECTION"
if-shell 'test -n "$SSH_CONNECTION"' 'set @ssh_connection "yes"'
set-hook -g session-created[100] 'source-file ~/.tmux.conf'
```

Only then I'll have a session-specific option `#{@ssh_connection}`. Then I use this in the custom module:

```
show_ssh() {
  text="$(  get_tmux_option "@catppuccin_ssh_text"  "#{@ssh_connection}" )"
  if [ -n "$text" ]; then
    # ...
    module=$( build_status_module "$index" "$icon" "$color" "" )
    echo "$module"
  fi
}
```

It does reflect the ssh state properly. Yet the problem I described still persists.

**Reply by vdbe** [reply]
Have you already looked at how other plugins do it for example:
- https://github.com/soyuka/tmux-current-pane-hostname/blob/6bb3c95250f8120d8b072f46a807d2678ecbc97c/scripts/shared.sh#L98
- https://github.com/obxhdx/tmux-ssh-status/blob/5d786c676f1bad6bdc1d9be5074859ba7e00427d/scripts/update_ssh_status.sh#L10

**Reply by vdbe** [reply]
```bash
  text="$(  get_tmux_option "@catppuccin_ssh_text"  "#{@ssh_connection}" )"
  if [ -n "$text" ]; then
    # ...
    module=$( build_status_module "$index" "$icon" "$color" "" )
    echo
  fi
```
I think this code does not do what you expect.
`text` will be the value of returned by `get_tmux_option` which is the result of `tmux show-option -g "@catppuccin_ssh_text"`, if `@catppuccin_ssh_text` is _not_ set it will return the default specified by `$2` in this case `#{@ssh_connection}` as a _string_.

**Reply by twio142** [reply]
Both of the two approaches provide ssh info in a pane that is **connecting to a ssh client**, but what I aim at is to show **whether a session was logged in through ssh**. That is to tell between 1) when I'm logging in to my machine remotely through ssh, and 2) when I'm using it directly at home.

**Reply by vdbe** [reply]
Ah I ran into a similar issue with gpg and ssh agent forwarding over ssh with tmux on the remote. Never got it fully working (also tried update-environment) maybe someone has it working for that? 

Sorry that I can't help further, good luck.

**Reply by twio142** [reply]
My approach does work! Just need to use a hook to source `~/.tmux.conf` once more on `session-created`.

My only problem is with the status separators that can't be hidden…

**Reply by vdbe** [reply]
Can you add a screenshot then because I have no clue what status separators you are talking about.
`$status_left_separator` and `$status_right_separator` are both added in [build_status_module](https://github.com/catppuccin/tmux/blob/408c02ccf44d0a59a7a63ce2b65c5c29982c5c0e/builder/module_builder.sh#L3-L44) so if the variables are correct and you rerun the plugin (since that if is only ran during configuration not on `status-interval`) there should be no left/right separator.

**Reply by twio142** [reply]
This is when I open a remote tmux session (ssh to my own machine) from a local tmux session:

![image](https://github.com/user-attachments/assets/2672d7e0-80c9-4842-a066-f7c8f0849e44)

You can see the overlapped separators in the lower status bar. It creates an extra blank space there.

If I remove `ssh` from `@catppuccin_status_modules_right`:

![image](https://github.com/user-attachments/assets/5130211a-49f4-4d94-9835-2a6df71b0163)

And my custom module:

```
show_ssh() {
  local index icon color text module

  index=$1
  icon="$(  get_tmux_option "@catppuccin_ssh_icon"  "#{?@ssh_connection,􀌌,}"           )"
  color="$( get_tmux_option "@catppuccin_ssh_color" "#{?@ssh_connection,$thm_magenta,}" )"
  text="$(  get_tmux_option "@catppuccin_ssh_text"  "#{@ssh_connection}" )"

  [ -z $text ] ||
    module=$( build_status_module "$index" "$icon" "$color" "" )

  echo "$module"
}
```

**Reply by vdbe** [reply]
Have not tested it myself but on first look I think the extra space is from the empty echo. `echo ""` will print a new line char. Try setting the echo in an if statement with the `module= ... build_status_module ...`.

To have it completely remove the module you want `show_ssh` to print nothing to stdout

**Reply by twio142** [reply]
Hmmm tried this and the problem persists:

```
show_ssh() {
  # ...
  if [ -n $text ]; then
    module=$( build_status_module "$index" "$icon" "$color" "" )
    echo "$module"
  fi
}
```

It makes no sense to me neither.
Never mind, It's probably not worth any more effort.
Thanks for the quick response though! :)
